server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java

```diff
@@ -0,0 +1,482 @@
+
+                    @Override
+                    public void visit(IntsRef ref) {
+                        for (int i = 0; i < ref.length; i++) {
+                            adder.add(ref.ints[ref.offset + i]);
+                        }
+                    }
+
+                    @Override
+                    public void visit(int docID, byte[] packedValue) {
+                        if (matches(packedValue)) {
+                            visit(docID);
+                        }
+                    }
+
+                    @Override
+                    public void visit(DocIdSetIterator iterator, byte[] packedValue) throws IOException {
+                        if (matches(packedValue)) {
+                            adder.add(iterator);
+                        }
+                    }
+
+                    @Override
+                    public PointValues.Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
+                        return relate(minPackedValue, maxPackedValue);
+                    }
+                };
+            }
+
+            private boolean checkValidPointValues(PointValues values) throws IOException {
+                if (values == null) {
+                    // No docs in this segment/field indexed any points
+                    return false;
+                }
+
+                if (values.getNumIndexDimensions() != pointRangeQuery.getNumDims()) {
+                    throw new IllegalArgumentException(
+                        "field=\""
+                            + pointRangeQuery.getField()
+                            + "\" was indexed with numIndexDimensions="
+                            + values.getNumIndexDimensions()
+                            + " but this query has numDims="
+                            + pointRangeQuery.getNumDims()
+                    );
+                }
+                if (pointRangeQuery.getBytesPerDim() != values.getBytesPerDimension()) {
+                    throw new IllegalArgumentException(
+                        "field=\""
+                            + pointRangeQuery.getField()
+                            + "\" was indexed with bytesPerDim="
+                            + values.getBytesPerDimension()
+                            + " but this query has bytesPerDim="
+                            + pointRangeQuery.getBytesPerDim()
+                    );
+                }
+                return true;
+            }
+
+            private void intersectLeft(PointValues.PointTree pointTree, PointValues.IntersectVisitor visitor) throws IOException {
+                intersectLeft(visitor, pointTree);
+                assert pointTree.moveToParent() == false;
+            }
+
+            private void intersectRight(PointValues.PointTree pointTree, PointValues.IntersectVisitor visitor) throws IOException {
+                intersectRight(visitor, pointTree);
+                assert pointTree.moveToParent() == false;
+            }
```

Comment:
(Also, if you are relying on the `moveToParent` side-effect, I believe it would be removed if assertions are disabled, like in a production deployment.)

Reply:
The assert essentially ensures that after we traverse left or right we come back to the root of the tree so `pointTree.moveToParent()` will try to go one level up but should fail if we've correctly traversed the tree. I don't mind removing the assertions completely and removing the overload, but this should help asserting correctness of the algorithm? wdyt